sklearn/preprocessing/imputation.py

```diff
@@ -374,3 +375,304 @@ def transform(self, X):
+
+        if np.any(row_has_missing):
+            neighbors = self._fitted_neighbors.kneighbors(
+                X[row_has_missing, :], n_neighbors=n_neighbors_new)
+
+            # Get row index, distance, and weights of donors
+            knn_distances, knn_row_index = neighbors
+            # Remove self from list of donors
+            if n_neighbors_new > self.n_neighbors:
+                row_index = np.arange(X.shape[0]).reshape((X.shape[0], 1))
+                row_index = row_index[row_has_missing, :]
+                not_duplicate_index = np.where(~(row_index == knn_row_index))
+                knn_row_index = knn_row_index[not_duplicate_index].reshape(
+                    (-1, self.n_neighbors))
+                knn_distances = knn_distances[not_duplicate_index].reshape(
+                    (-1, self.n_neighbors))
+            weights = _get_weights(knn_distances, self.weights)
+
+            # Vertically split sets of k-donor indices and repeat each set by
+            # missing count in the corresponding recipient row
+            knn_row_index = np.vsplit(knn_row_index, knn_row_index.shape[0])
+            row_repeats = row_total_missing[row_total_missing != 0]
+            knn_row_index = np.repeat(
+                knn_row_index, row_repeats, axis=0).ravel()
+
+            # Get column index of donors
+            row_missing_index, col_missing_index = np.where(mask)
+            knn_col_index = np.repeat(col_missing_index, self.n_neighbors)
+
+            # Calculate kNN score and impute
+            donors = fitted_X[
+                (knn_row_index, knn_col_index)].reshape((-1, self.n_neighbors))
+            donors_mask = _get_mask(donors, self.missing_values)
+            donors = np.ma.array(
+                donors, mask=donors_mask)
+            imputed = np.ma.average(donors, axis=1, weights=weights)
+            X[mask] = imputed.data
+            unimputed_index = np.where(
+                donors_mask.sum(axis=1) == self.n_neighbors)
+            if len(unimputed_index[0]) > 0:
+                unimputed_rows = row_missing_index[unimputed_index]
+                unimputed_cols = col_missing_index[unimputed_index]
+                X[(unimputed_rows, unimputed_cols)] = np.take(self.statistics_,
+                                                              unimputed_cols)
+
+        # Merge bad rows to X and mean impute their missing
+        if np.any(bad_rows):
+            bad_missing_index = np.where(_get_mask(X_bad, self.missing_values))
+            X_bad[bad_missing_index] = np.take(self.statistics_,
+                                               bad_missing_index[1])
+            X_merged = np.empty((n_rows_X, n_cols_X))
+            X_merged[bad_rows, :] = X_bad
+            X_merged[~bad_rows, :] = X
+            X = X_merged
+        return X
+
+    def fit_transform(self, X, y=None, **fit_params):
+        """Fit KNNImputer and impute all missing values in X. This method
```

Comment:
PEP257: the first line should be a standalone summary, with fuller description separated off by a blank line